crates/typst/src/math/attach.rs

```diff
@@ -52,31 +52,55 @@ pub struct AttachElem {
 impl LayoutMath for Packed<AttachElem> {
     #[typst_macros::time(name = "math.attach", span = self.span())]
     fn layout_math(&self, ctx: &mut MathContext, styles: StyleChain) -> SourceResult<()> {
-        type GetAttachment = fn(&AttachElem, styles: StyleChain) -> Option<Content>;
-
-        let layout_attachment =
-            |ctx: &mut MathContext, styles: StyleChain, getter: GetAttachment| {
-                getter(self, styles)
-                    .map(|elem| ctx.layout_into_fragment(&elem, styles))
-                    .transpose()
-            };
-
         let base = ctx.layout_into_fragment(self.base(), styles)?;
 
         let sup_style = style_for_superscript(styles);
-        let tl = layout_attachment(ctx, styles.chain(&sup_style), AttachElem::tl)?;
-        let tr = layout_attachment(ctx, styles.chain(&sup_style), AttachElem::tr)?;
-        let t = layout_attachment(ctx, styles.chain(&sup_style), AttachElem::t)?;
+        let sup_style_chain = styles.chain(&sup_style);
+        let tl = self.tl(sup_style_chain);
+        let (tr, primes) = {
+            let tr = self.tr(sup_style_chain);
+            if tr.as_ref().is_some_and(|content| content.is::<PrimesElem>()) {
+                (None, tr)
+            } else {
+                (tr, None)
+            }
+        };
+        let t = self.t(sup_style_chain);
 
         let sub_style = style_for_subscript(styles);
-        let bl = layout_attachment(ctx, styles.chain(&sub_style), AttachElem::bl)?;
-        let br = layout_attachment(ctx, styles.chain(&sub_style), AttachElem::br)?;
-        let b = layout_attachment(ctx, styles.chain(&sub_style), AttachElem::b)?;
+        let sub_style_chain = styles.chain(&sub_style);
+        let bl = self.bl(sub_style_chain);
+        let br = self.br(sub_style_chain);
+        let b = self.b(sub_style_chain);
 
         let limits = base.limits().active(styles);
-        let (t, tr) = if limits || tr.is_some() { (t, tr) } else { (None, t) };
+        let (t, mut tr) = if limits || tr.is_some() { (t, tr) } else { (None, t) };
+        if let Some(primes) = primes {
+            tr = tr.map_or(Some(primes.clone()), |tr| Some(primes + tr));
```

Comment:
should probably be `map_or_else`

Reply:
Hi - it turned out `map_or_else` needs even more `.clones()` to satisfy the borrow-checker
```
tr = tr.map_or_else(|| Some(primes.clone()), |tr| Some(primes.clone() + tr));
```
I also tried `match`, but it turned out it couldn't reduce `.clone()` calls either.

Reply:
I saw this and got nerd sniped :) The issue is that you convert `tr` and `primes` into separate, yet *exclusive* options above, which confuses the borrow checker since `tr` isn't actually a separate value than `primes`.

This is much cleaner (and makes the logic easier to understand) if we instead convert `primes` into a bool and then just do a match on `t` and `tr` with separate guards:

```diff
diff --git a/crates/typst/src/math/attach.rs b/crates/typst/src/math/attach.rs
index 8e679627..ad10c1ac 100644
--- a/crates/typst/src/math/attach.rs
+++ b/crates/typst/src/math/attach.rs
@@ -60,8 +60,2 @@ impl LayoutMath for Packed<AttachElem> {
-        let (tr, primes) = {
-            let tr = self.tr(sup_style_chain);
-            if tr.as_ref().is_some_and(|content| content.is::<PrimesElem>()) {
-                (None, tr)
-            } else {
-                (tr, None)
-            }
-        };
+        let tr = self.tr(sup_style_chain);
+        let primes = tr.as_ref().is_some_and(|content| content.is::<PrimesElem>());
@@ -77,4 +71,5 @@ impl LayoutMath for Packed<AttachElem> {
-        let (t, mut tr) = if limits || tr.is_some() { (t, tr) } else { (None, t) };
-        if let Some(primes) = primes {
-            tr = tr.map_or(Some(primes.clone()), |tr| Some(primes + tr));
-        }
+        let (t, tr) = match (t, tr) {
+            (Some(t), Some(tr)) if primes && !limits => (None, Some(tr + t)),
+            (Some(t), None) if !limits => (None, Some(t)),
+            (t, tr) => (t, tr),
+        };
```

Probably should also update the `b` and `br` rules for symmetry (and clarity), unless you prefer leaving it as one line.

Reply:
Thanks! It works and I added a new commit :)